script.service.kodi.callbacks/default.py

```diff
@@ -0,0 +1,316 @@
+    paths.append(os.path.join(paths[0], 'lib'))
+    paths.append(os.path.join(paths[1], 'usertasks'))
+    for path in paths:
+        if not os.path.isdir(path):
+            try:
+                os.mkdir(path)
+            except OSError:
+                pass
+    for path in paths[1:]:
+        fn = os.path.join(path, '__init__.py')
+        if not os.path.isfile(fn):
+            try:
+                with open(fn, mode='w') as f:
+                    f.writelines('')
+            except (OSError, IOError):
+                pass
+
+
+class MainMonitor(xbmc.Monitor):
+    def __init__(self, dispatcher, publishers):
+        super(MainMonitor, self).__init__()
+        self.dispatcher = dispatcher
+        self.publishers = publishers
+
+    def onSettingsChanged(self):
+        log(msg=_('Settings change detected - attempting to restart'))
+        for p in self.publishers:
+            p.abort(0.525)
+        self.dispatcher.abort(0.25)
+        start()
+
+
+def start():
+    global log
+    settings = Settings()
+    settings.getSettings()
+    kl = KodiLogger()
+    if settings.general['elevate_loglevel'] is True:
+        kl.setLogLevel(xbmc.LOGNOTICE)
+    else:
+        kl.setLogLevel(xbmc.LOGDEBUG)
+    log = kl.log
+    log(msg=_('Settings read'))
+    dispatcher = PubSub_Threaded.Dispatcher(interval=settings.general['TaskFreq'], sleepfxn=xbmc.sleep)
+    log(msg=_('Dispatcher initialized'))
+
+    subscriberfactory = SubscriberFactory(settings, kl)
+    subscribers = subscriberfactory.createSubscribers()
+    for subscriber in subscribers:
+        dispatcher.addSubscriber(subscriber)
+    publisherfactory = PublisherFactory(settings, subscriberfactory.topics, dispatcher, kl, debug)
+    publisherfactory.createPublishers()
+    publishers = publisherfactory.ipublishers
+
+    dispatcher.start()
+    log(msg=_('Dispatcher started'))
+
+    for p in publishers:
+        try:
+            p.start()
+        except threading.ThreadError:
+            raise
+    log(msg=_('Publisher(s) started'))
+    return dispatcher, publishers
+
+
+def main():
+    xbmc.log(msg=_('$$$ [kodi.callbacks] - Staring kodi.callbacks ver: %s') % str(__version__), level=xbmc.LOGNOTICE)
+    if branch != 'master':
+        xbmcaddon.Addon().setSetting('installed branch', branch)
+    dispatcher, publishers = start()
+    dispatcher.q_message(PubSub_Threaded.Message(PubSub_Threaded.Topic('onStartup')))
+    monitor = MainMonitor(dispatcher, publishers)
+    log(msg=_('Entering wait loop'))
+    monitor.waitForAbort()
+
+    # Shutdown tasks
+    dispatcher.q_message(PubSub_Threaded.Message(PubSub_Threaded.Topic('onShutdown'), pid=os.getpid()))
+    xbmc.sleep(500)
+    log(msg=_('Shutdown started'))
+    for p in publishers:
+        try:
+            p.abort()
+        except threading.ThreadError as e:
+            log(msg=_('Error aborting: %s - Error: %s') % (str(p), str(e)))
+    dispatcher.abort()
+    xbmc.sleep(1000)
```

Comment:
Hopefully I can explain the rationale for this so that it is seen that all I am trying to do is catch situations where the threads don't abort. Except in running tests, I have yet to see that this code has ever run on a 'live' Kodi installation in my own or anyone else's logs. So here it goes:

Problem:
In my own logs I see other addons not exit properly and then the dreaded 'xxx has not responded in 5 seconds, let's kill it'. I think, if it were my addon that doesn't respond, I would be embarrassed. I think to myself well you have tested it and it doesn't do that. But then I realize that my testing consists of my own high-ish powered windows machine, two VM's running LinuxMint and OSX 10.11 and a bluestacks Android Emulator. What if it hangs on someone's Rpi running OpenElec? Just let the embarrassment happen? Or try to plan for it?

All of the publishers are running loops in their own separate threads. The dispatcher is in a loop it's own thread. Each of these runs their own code and then sleeps and then checks if an abort has been set in it's own threading.Event(). The amount of time it takes the code portion to run is going to depend on many things so I don't truly know how often each loop is checking for it's own abort event to be set. In addition each of the tasks runs in its own thread and the end user may have written code that takes an unknown amount of time to run. However, I do know that I have five seconds before big brother Kodi shuts me down. So maybe in that five seconds I can catch situations where things have not gone as planned and either fix them or at least write into the log what happened so that I can change the code. So I set up tests where either the publisher, the dispatcher doesn't respond to the abort or if the task is long running.

So on line 137 I publish the onShutdown message and wait an arbitrary 500ms for it to do it's thing and then I send the abort signals to the publishers and the dispatcher (it's really the dispatcher that I don't want to kill before the onShutdown event completes).
On line 145 I give the publishers and dispatcher an arbitrary 1000ms (because of the unknown duration of the loops they are running) before I start looking to see if _any_ of the threads that I own haven't exited (lines 146-159). The sleep in line 145 is there because during _testing_ I found that occasionally the message doesn't make it into the log if the thread aborts successfully in the remaining code. Perhaps the 525ms in thread.abort() is too long, Note that only works if it is one of my own threads that subclassed threading.Thread. I should also catch the exception where the thread does not implement abort. I then log if I was able to terminate the threads. So at least if my addon doesn't exit as planned the log can guide me back to identify where the bad actor is.

Again, I have yet to see this code run on a live install in anyone's logs. Anyway I hope this explains why the code is there.